kimchi/src/circuits/polynomials/foreign_mul/gate.rs

```diff
@@ -0,0 +1,528 @@
+                wires: wires[3],
+                coeffs: vec![],
+            },
+            /* Input: b */
+            CircuitGate {
+                typ: GateType::ForeignMul0,
+                wires: wires[4],
+                coeffs: vec![],
+            },
+            CircuitGate {
+                typ: GateType::ForeignMul0,
+                wires: wires[5],
+                coeffs: vec![],
+            },
+            CircuitGate {
+                typ: GateType::ForeignMul1,
+                wires: wires[6],
+                coeffs: vec![],
+            },
+            CircuitGate {
+                typ: GateType::ForeignMul2,
+                wires: wires[7],
+                coeffs: vec![],
+            },
+        ]
+    }
+
+    // Verify the foreign field multiplication circuit gate on a given row
+    pub fn verify_foreign_mul(
+        &self,
+        row: usize,
+        witness: &[Vec<F>; COLUMNS],
+        cs: &ConstraintSystem<F>,
+    ) -> Result<(), String> {
+        // Witness row shorthands
+        let curr: [F; COLUMNS] = array_init(|i| witness[i][row]);
+        let next: [F; COLUMNS] = array_init(|i| witness[i][row + 1]);
+
+        // Columns types that need to be evaluated for this gate to work
+        let evaluated_cols = {
+            let mut h = std::collections::HashSet::new();
+            // Witness columns
+            for i in 0..COLUMNS {
+                h.insert(Column::Witness(i));
+            }
+            // Index column for each foreign field multiplication circuit gate
+            h.extend(
+                circuit_gates()
+                    .iter()
+                    .map(|gate_type| Column::Index(*gate_type)),
+            );
+            h
+        };
+
+        // Setup powers of alpha
+        let mut alphas = Alphas::<F>::default();
+        alphas.register(
+            ArgumentType::Gate(self.typ),
+            circuit_gate_constraint_count::<F>(self.typ),
+        );
+
+        // Get constraints for this circuit gate
+        let constraints = circuit_gate_constraints(self.typ, &alphas);
+
+        // Linearize
+        let linearized = constraints.linearize(evaluated_cols).unwrap();
+
+        // Setup proof evaluations
+        let rng = &mut StdRng::from_seed([0u8; 32]);
+        let mut eval = |witness| ProofEvaluations {
+            w: witness,
+            z: F::rand(rng),
+            s: array_init(|_| F::rand(rng)),
+            generic_selector: F::zero(),
+            poseidon_selector: F::zero(),
+            lookup: None,
+            foreign_mul_selector: match self.typ {
+                // Force this circuit gate on
+                GateType::ForeignMul0 => vec![F::one(), F::zero(), F::zero()],
+                GateType::ForeignMul1 => vec![F::zero(), F::one(), F::zero()],
+                GateType::ForeignMul2 => vec![F::zero(), F::zero(), F::one()],
+                _ => panic!("Invalid gate type {:?}", self.typ),
+            },
+        };
```

Comment:
we really need a `for_testing` for `ProofEvaluations` instead of rewriting this code all the time :)

Reply:
Spoke with Matthew, this was a nit and will be done in next iteration's PR.